### PR TITLE
Rewritten Win32 music driver

### PR DESCRIPTION
**Pull request by nielsmh**
Just the new Win32 music driver extracted from PR #6749.

Implements MIDI file decoding directly and does not depend on MCI or DirectSound. (Both are troublesome in modern Windows versions.)

### CODE REVIEW

File: src/music/midifile.cpp [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+	MIDIST_SMF_ESCAPE = 0xF7, ///< only occurs in SMF data
+	MIDIST_TC_QFRAME = 0xF1,
+	MIDIST_SONGPOSPTR = 0xF2,
+	MIDIST_SONGSEL = 0xF3,
+	MIDIST_TUNEREQ = 0xF6,
+	MIDIST_RT_CLOCK = 0xF8,
+	MIDIST_SYSRESET = 0xFF, ///< only occurs in realtime data
+	MIDIST_SMF_META = 0xFF, ///< only occurs in SMF data
+};
+enum MidiController {
```
Comment (michicc) [inline src/music/midifile.cpp]: Empty line? And maybe some Doxygen at least for enums itself (above as well)?

File: src/music/midifile.cpp [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+	}
+};
+
+static bool ReadTrackChunk(FILE *file, MidiFile &target)
+{
+	byte buf[4];
+
+	const byte magic[] = { 'M', 'T', 'r', 'k' };
+	if (fread(buf, sizeof(magic), 1, file) != 1) {
+		return false;
```
Comment (michicc) [inline src/music/midifile.cpp:155]: Could also be formatted as a single-line `if`. Personally I'd prefer that, but that's very subjective and a matter of taste I guess.

File: src/music/win32_m.cpp [hunk 3]
```diff
@@ -15,170 +15,361 @@
...
+	remaining -= msg_end - msg_start;
+	msg_start = msg_end;
 }
 
-static MCIERROR CDECL MidiSendCommand(const TCHAR *cmd, ...)
+void CALLBACK TimerCallback(UINT uTimerID, UINT, DWORD_PTR dwUser, DWORD_PTR, DWORD_PTR)
 {
-	va_list va;
-	TCHAR buf[512];
+	if (TryEnterCriticalSection(&_midi.lock)) {
```
Comment (michicc) [inline src/music/win32_m.cpp]: What's the effect of not being able to acquire the lock?

File: src/music/win32_m.cpp [hunk 4]
```diff
@@ -15,170 +15,361 @@
...
+			byte status = data[0];
+			if (status & 0x80) {
+				last_status = status;
+				data++;
+				remaining--;
+			} else {
+				status = last_status;
+			}
+			switch (status & 0xF0) {
+				case 0xC0: case 0xD0:
```
Comment (michicc) [inline src/music/win32_m.cpp]: Are these values from the `MidiStatus` enum? If they are, using the enum here would be better than magic numbers.

File: src/music/win32_m.cpp [hunk 5]
```diff
@@ -339,10 +339,14 @@ const char *MusicDriver_Win32::Start(const char * const *parm)
...
 		static byte gm_enable_sysex[] = { 0xF0, 0x7E, 0x00, 0x09, 0x01, 0xF7 };
-		byte *ptr = gm_enable_sysex;
-		size_t len = sizeof(gm_enable_sysex);
-		TransmitSysex(ptr, len);
+		/* Roland-specific reverb room control, used by the original game */
+		static byte roland_reverb_sysex[] = { 0xF0, 0x41, 0x10, 0x42, 0x12, 0x40, 0x01, 0x30, 0x02, 0x04, 0x00, 0x40, 0x40, 0x00, 0x00, 0x09, 0xF7 };
+		byte *ptr;
+		size_t len;
+		TransmitSysex((ptr = gm_enable_sysex), (len = sizeof(gm_enable_sysex)));
+		TransmitSysex((ptr = roland_reverb_sysex), (len = sizeof(roland_reverb_sysex)));
```
Comment (michicc) [inline src/music/win32_m.cpp]: I'm not sure I really like that syntax...

File: src/music/win32_m.cpp (hunk 3 above)
Comment (nielsmh) [reply src/music/win32_m.cpp]: Good point, I'll write some comments there. The function is the realtime playback service routine, so it just attempts to get a lock on the playback status data, and skips checking it if they're currently locked.

File: src/music/win32_m.cpp (hunk 4 above)
Comment (nielsmh) [reply src/music/win32_m.cpp]: I considered that, but then I'd have to put them into a header file. I'm not sure if I want to introduce a `midifile_internal.h`, it feels overkill.

File: src/music/midifile.cpp (hunk 2 above)
Comment (nielsmh) [reply src/music/midifile.cpp:155]: Originally I had a mix of single-line and braced early returns like this. I thought it made the early return cases stand out more in a good way to format them braced.

File: src/music/win32_m.cpp (hunk 5 above)
Comment (nielsmh) [reply src/music/win32_m.cpp]: Oh, someone discovered it ;) Yeah may as well introduce another function to make it cleaner.

File: src/music/win32_m.cpp (hunk 4 above)
Comment (nielsmh) [reply src/music/win32_m.cpp]: Working on moving this now. Also found a couple mistakes in the declared enums.

**Review by michicc — APPROVED** [review]

**Pull request merged**